### ISSUE TITLE
how to cancel a running solver?

### ISSUE THREAD

**Issue by jwaldmann** [niklasso/minisat-haskell-bindings#1]
Hi. I have an application where I start several minisat instances (on different CNFs) in parallel, and as soon as one of them reports "solved", I want to cancel the others. How can I do this via the Haskell API? 

As it is now, it does not seem possible, cf. https://github.com/simonmar/async/issues/8

I guess a work-around would be to make Minisat run in chunks, and return to Haskell after each 10^6 propagations or similar. So the "solver main loop" can be written in Haskell land, and can be interrupted.

**Comment by niklasso** [comment]
There is support for this in MiniSat but it's not exposed in the haskell interface. Here's what you can do in c++:
- For solving call solveLimited(). This can return undefined if it runs out of a given budget, but even better in your case, it can also return after an asynchronous interrupt call was made by a separate thread.
- You can have a separate thread that waits for a timeout and then calls interrupt(). SolveLimited() will the return undefined to indicate this.
- To continue solving clearInterrupt() first have to be called or the next call to solveLimited() will just terminate directly.

This should work, and you'd not have to periodically poll the solver.

Now, adding these feature shouldn't be too hard but I'm not sure I'll have time for that any time soon. Would you like to have a go at this? What needs to be done is: Add interrupt()/clearInterrupt() to the c-interface. Add solveLimited/interrupt/clearInterrupt to the haskell interface. There might be some choice here as to what makes most sense for a haskell interface. Currently we made the intentional choice to have the haskell-interface very low-level, so just directly mirroring the c interface is probably fine (the idea is that any intelligence can be added as a layer on top of this if needed).

**Comment by jwaldmann** [comment]
OK, sounds easy enough. I want to make sure I understood:
- in minisat-c-bindings/minisat.h, it would look like  void minisat_interrupt (minisat_solver *s)
- I call this from a separate Haskell thread, and then the call of minisat_solve_limited returns (and consequently, the Haskell thread that was waiting for the result, can continue, with "undefined" result)

having a low-level Haskell interface is fine, I hide this behind my library anyway

**Comment by niklasso** [comment]
Yes, that should should work. I seem to recall that there used to be issues with the whole haskell runtime being suspended during FFI calls, but you probably know more than me about what the current state of that is. Could it be that we need to change the "import" declarations of the functions? I.e. is it fine to let them be "unsafe" or should they be something else to support multi-threading properly?

**Comment by jwaldmann** [comment]
I declared "#safe minisat_limited_solve_commit" because I saw "#safe minisat_solve_commit".

This text http://blog.ezyang.com/2010/07/safety-first-ffi-and-threading/ argues that most everything should be declared safe.

So I think it's fine.

**Comment by niklasso** [comment]
Yes, I've read about this in the past but never found it very clear
what/how one should choose between safe/unsafe. If I'm reading that blog
post + comments correctly, is it the case that things that are constant
time / fast are ok to declare "unsafe" and everything else should be
"safe"? Then one or two more should probably be declared safe I think.

Niklas

**Comment by jwaldmann** [comment]
cf. remark here: https://github.com/ghc/ghc/blob/master/rts/Schedule.c#L2147

I checked with -ddump-asm that the compiler indeed produces suspendThread/resumeThread around minisat_limited_solve_commit, and it doesn't around, e.g., minisat_newLit.

So I guess any FFI function that's not O(1) should be declared "safe"?

**Comment by niklasso** [comment]
Yes, sounds good, and I'm now starting to understand :) How about average
case O(1)? To be precise that's what minisat_newLit actually is, because it
is somtimes doubling a vector of variable-information. I suspect that's
reasonably safe to declare "unsafe".

/Niklas

**Comment by jwaldmann** [comment]
Sounds OK to me, but obviously you have better knowledge of minisat internals and runtimes of functions. 

From an application programmer's viewpoint it would be preferrable that constructing a formula (newLit) and accessing a model has as little overhead as possible (that is, should be "unsafe" in the ghc FFI sense).
